plutus-tx/src/PlutusTx/Builtins.hs

```diff
@@ -184,22 +187,23 @@ verifyEd25519Signature
     -> BuiltinByteString  -- ^ Message    (arbirtary length)
     -> BuiltinByteString  -- ^ Signature  (64 bytes)
     -> Bool
-verifyEd25519Signature pubKey message signature = fromBuiltin (BI.verifyEd25519Signature pubKey message signature)
+verifyEd25519Signature pubKey message signature =
+    fromOpaque (BI.verifyEd25519Signature pubKey message signature)
 
 {-# INLINABLE equalsByteString #-}
 -- | Check if two 'ByteString's are equal.
 equalsByteString :: BuiltinByteString -> BuiltinByteString -> Bool
-equalsByteString x y = fromBuiltin (BI.equalsByteString x y)
+equalsByteString x y = fromOpaque (BI.equalsByteString x y)
```

Comment:
I have to admit that I thought I understood the difference between `HasBuiltin` and `HasOpaque` but in this case I have no idea what to use and if `fromOpaque` should be used instead of the `fromBuiltin` 😕 

If "opaque" types are not-pattern-matchable does it mean that we convert from them using `fromOpaque` with the sole intention to then pattern match?

Reply:
I am also confused, but from what I understood is that

`from/toOpaque` is supposed to be sth like Coercible but for **coercing** (0-cost) between Haskell and Tx universe (only that part that coincides),

whereas `from/toBuiltin` is for **converting** (thus, potentially non-zero cost) from a Haskell ADT (e.g. List) to a Tx builtin (primitive) type (e.g. BuiltinList). A List adt is a container of any `a`, whereas a BuiltinList can only contain `forall a in uni`.

Reply:
Actually I take it back I am now more confused.

Reply:
> I have to admit that I thought I understood the difference between HasBuiltin and HasOpaque but in this case I have no idea what to use and if fromOpaque should be used instead of the fromBuiltin 😕

The Note literally instructs you to use `fromOpaque` in that case:

> This is what the 'HasToOpaque' and 'HasFromOpaque' classes do. They let us write wrappers for
> builtins relatively consistently by just calling 'toOpaque' on their arguments and 'fromOpaque' on
> the result.

> If "opaque" types are not-pattern-matchable does it mean that we convert from them using fromOpaque with the sole intention to then pattern match?

With the intention to be able to do with them anything you can do with a regular Plutus Tx data type, in particular pattern matching and using library functions.